bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java

```diff
@@ -215,86 +217,111 @@ public boolean exists(String subscriptionId) {
-                    pageId, sitemapName, callbacks.size());
+            logger.debug("Subscription {} changed to {} ({} active subscriptions}", subscriptionId, scopeIdentifier,
+                    callbacks.size());
         } else {
             throw new IllegalArgumentException("Subscription " + subscriptionId + " does not exist!");
         }
     }
 
-    private void addCallbackToListener(String sitemapName, String pageId, SitemapSubscriptionCallback callback) {
-        ListenerRecord listener = pageChangeListeners.computeIfAbsent(getValue(sitemapName, pageId), v -> {
-            PageChangeListener newListener = new PageChangeListener(sitemapName, pageId, itemUIRegistry,
+    private void addCallbackToListener(String sitemapName, @Nullable String pageId,
+            SitemapSubscriptionCallback callback) {
+        String sitemapWithPageId = getScopeIdentifier(sitemapName, pageId);
+        ListenerRecord listener = pageChangeListeners.computeIfAbsent(sitemapWithPageId, v -> {
+            WidgetsChangeListener newListener = new WidgetsChangeListener(sitemapName, pageId, itemUIRegistry,
                     collectWidgets(sitemapName, pageId));
             ServiceRegistration<?> registration = bundleContext.registerService(EventSubscriber.class.getName(),
                     newListener, null);
             return new ListenerRecord(newListener, registration);
         });
-        listener.pageChangeListener().addCallback(callback);
+        listener.widgetsChangeListener().addCallback(callback);
     }
 
-    private EList<Widget> collectWidgets(String sitemapName, String pageId) {
+    public EList<Widget> collectWidgets(String sitemapName, @Nullable String pageId) {
         EList<Widget> widgets = new BasicEList<>();
 
         Sitemap sitemap = getSitemap(sitemapName);
-        if (sitemap != null) {
-            if (pageId.equals(sitemap.getName())) {
-                widgets = itemUIRegistry.getChildren(sitemap);
-            } else {
-                Widget pageWidget = itemUIRegistry.getWidget(sitemap, pageId);
-                if (pageWidget instanceof LinkableWidget widget) {
-                    widgets = itemUIRegistry.getChildren(widget);
-                    // We add the page widget. It will help any UI to update the page title.
-                    widgets.add(pageWidget);
+        if (sitemap == null) {
+            // no sitemap found with the given name
+            return widgets;
+        }
+
+        if (pageId != null && !pageId.equals(sitemap.getName())) {
+            // subscribing to subpage of sitemap --> get all widgets from that page
+            Widget pageWidget = itemUIRegistry.getWidget(sitemap, pageId);
+            if (pageWidget instanceof LinkableWidget widget) {
+                widgets = itemUIRegistry.getChildren(widget);
+                // We add the page widget. It will help any UI to update the page title.
+                widgets.add(pageWidget);
```

Comment:
In the current implementation, `pageWidget`was not considered by `blockUnlessChangeOccurs`.

I am not sure if it was an error or it was not an error what could be the impact of considering it now.
WDYT ?

Reply:
I think both implementations for collecting widgets are meant to do the same and omitting the page in one of them was the typical consequence of code duplication. I will check what that causes with the tests though.

Reply:
I just hope it was not done deliberately.

Reply:
I will take a look at the history of the code before resolving this conversation

Reply:
Is `blockUntilChangeOccurs` only used when long polling mode is used ? Sorry, I do not remember.

Reply:
Yes

Reply:
In this case, this will not impact Basic UI when used with any modern browser (meaning supporting SSE).
But not sure about Android app and iOS app, even if I know that the Android app is also using a SSE connection.

By the way, as previously said, the change probably makes sense.

Reply:
Seems like both places have been there since the code was added to the repository in 2019 (a37cceab), so I can´t judge. But the effect will only be a little more frequent updates, that occur also when the parent page itself changes, so I agree that it will probably not cause any trouble.